---
title: "class14"
author: "Bayah Essayem (A17303992)"
date: 02-20-2025
format: pdf
toc: TRUE
---

## Background

The data for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

> Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.

## Data Import 

```{r}
library(DESeq2)
counts <- read.csv("GSE37704_featurecounts.csv", row.names = 1)
colData <- read.csv("GSE37704_metadata.csv")
```


## Inspect and tidy data

Does the 'counts' column match the 'colData' rows? No.

```{r}
head(counts)
```

```{r}
head(colData)
```


```{r}
colData$id
```

```{r}
colnames(counts)
```

> Q1. Complete the code below to remove the troublesome first column from countData

The fix here is to remove the first "length" column from counts:
```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(counts[,-1])
head(counts)
```


Check for matching countData and colData

```{r}
colnames(countData) == colData$id
```

> Q. How many genes in total? 19808

```{r}
nrow(countData)
```
> Q2. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns). 

```{r}
head(rowSums(countData))
```


> Q. Filter to remove zero count genes (rows where there are zero counts in all columns). How many genes are left? 15975

```{r}
to.keep.inds <- rowSums(countData) > 0
head(countData)
```

```{r}
new.counts <- countData[to.keep.inds,]
```

```{r}
nrow(new.counts)
```


## Setup for DESeq

```{r}
#| message: FALSE
library(DESeq2)
```


Setup input object for DESeq

```{r}
dds <- DESeqDataSetFromMatrix(countData = new.counts,
                              colData = colData,
                              design =~condition)
```

## Run DESeq(dds)

```{r}
dds <- DESeq(dds)
res <- results(dds)
```

```{r}
head(res)
```


## Volcano plot of results

```{r}
library(ggplot2)
```



```{r}
ggplot(res) +
  aes(res$log2FoldChange, -log(res$padj)) +
  geom_point(alpha = 3, col = "pink", shape = 1) 
```


## Gene Annotation

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```

```{r}
columns(org.Hs.eg.db)
```

Add gene SYMBOL and ENTREZID

```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=rownames(res), 
                    keytype="ENSEMBL",
                    column="ENTREZID")

res$symbol = mapIds(org.Hs.eg.db,
                    keys=rownames(res),
                    keytype="ENSEMBL",
                    column="SYMBOL")

```

```{r}
head(res)
```


## Pathway Analysis

```{r}
library(gage)
library(gageData)
library(pathview)
```

Input vector for 'gage()'

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Load up the KEGG gene sets

```{r}
data("kegg.sets.hs")
```

Run pathway analysis with KEGG

```{r}
keggres = gage(foldchanges, gsets = kegg.sets.hs)
```


```{r}
head(keggres$less, 3)
```


## Cell Cycle Figure - include at least 3

```{r}
pathview(foldchanges, pathway.id = "hsa04110")
```

![](hsa04110.pathview.png)


DNA Replication Figure

## Gene Ontology Analysis

Run pathway analysis with GO

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO

gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage (foldchanges, gsets = gobpsets, same.dir = TRUE)

head(gobpres$less)
```


